### DISCUSSION TITLE
[창훈] assignment 기능구현

### DISCUSSION THREAD

**Discussion by anotheranotherhoon** [Wanted-07-team-9/pre-onboarding-7th-3-1-9#14] [기능구현]
## 사용 Tools
- React
- TypeScript
- styled-components
## ✏ 토의 내용 및 선정

[🖍 1차 토의결과](https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/discussions/1)  
[🖍 2차 토의결과](https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/discussions/1)

[🥇 선정 결과](https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/discussions/1)

## **📝 디렉토리 구조**

```
   📂src
   ┗ 📂api
   ┗ 📂assets
   ┗ 📂components
   ┗ 📂interface
   ┗ 📂container
   ┣📄App.tsx
   ┗📄index.tsx

```

## 질환명 검색시 API 호출 통해서 검색어 추천 기능 구현 ✔️ 
![스크린샷 2022-11-11 오전 1 17 56](https://user-images.githubusercontent.com/77476348/201149385-c17f6ef6-b525-433d-84ad-e693c3347c60.png)

## 사용자가 입력한 텍스트와 일치하는 부분 볼드처리 ✔️ 
https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/0894d6bb6af0ef6c8e43793e3c52a74d91d37c01/src/components/AutoSearchData/index.tsx#L5-L18

키워드로 받아온 추천검색어 데이터를 키워드로 split한다. split을 통해 반환된 배열을 map돌고 (배열의 아이템 한번, 키워드 한번) 순서대로 작성되도록 한다. split을 통해 반환된 배열의 마지막아이템의 뒤에는 키워드가 오지 않으므로 그 경우에는 아래의 코드를 실행하지 않도록한다.
```jsx
{i !== textArray.length - 1 && <span className="font_bold">{keyword}</span>}
```

## Q1. 캐싱 구현 방법

>Q. 왜 cacheStorage인가 ?
>
>A. cacheStorage는 localStorage,sessionStorage와 비교하여 훨씬 많은 데이터를 저장할 수 있다.  적어도 수백 MB, 경우에 따라 수 GB 이상까지도 될 수 있습니다

cacheStorage에 특정 이름(sick)으로 cach를 open한다

- case 1. cacheStorage에 검색 결과가 없을 경우 
서버주소로 get요청을 보내서 받아온 데이터를 cacheStorage에  sickNm_like_keyword 항목을 생성하고 응답 결과를 저장한다.

- case 1. cacheStorage에 검색 결과가 존재할 경우 
캐시에서 항목을 찾기 위해 match 메서드를 사용하고 일치하는 항목이 발견되면 Promise를 반환한다. 반환한 프로미스를 리턴해준다.

## Q2. 입력마다 API 호출하지 않도록 API 호출 횟수를 줄이는 방법

https://github.com/Wanted-07-team-9/pre-onboarding-7th-3-1-9/blob/bf32e970c26ef66f054171de698d1044774d42d2/src/App.tsx#L40-L50

만약 setTimeout없이 계속 keyword가 있을때마다 getSick()함수를 호출한다면 예를들어 담낭를 검색한다면
ㄷ ㅏ ㅁ ㄴ ㅏ ㅇ 총 6번을 호출 하므로 딜레이(0.2초)를 두어  API 호출 횟수를 줄였다.

## Q3. 키보드로 추천 검색어로 이동하는 방법

input 요소에 onKeyUp 이벤트에 실행될 함수를   바인딩해준다. e.key 가 'ArrowUp'(키보드 위쪽 화살표)  또는 'ArrowDown'(키보드 아래쪽 화살표)의 경우로 나누어준다. 
maxIndex변수에  순회하는 추천검색어 배열 (keyItems)의 갯수 -1 을 할당한다. 그 이유는 Index로 접근하기 때문이다. 
useState로 선택된 추천 검색어의 Index를 저장한다. 초기값을 -1로 하여 위, 아래 화살표를 누르지 않을 경우 추천검색어들이 선택되지 않은 상태로 해준다.

- case 1. 'ArrowUp' 일 경우 
  - 최초에  'ArrowUp' 키를 누른 경우 가장 최하단 추천 검색어로 이동해야 하기 때문에 selectedIndex의 값을 maxIndex값으로 바꿔준다. 
  - 그 이외의 경우 자신보다 위의 인덱스 값이 있으므로 selectIndex 값을 이전 값에서 -1 해준다.

- case 1. 'ArrowDown' 일 경우 
  - 추천 검색어의 마지막에서  'ArrowDown' 키를 누른 경우 아무것도 선택하지 않은 경우(-1)로 만든다.
  - 그 이외의 경우 자신보다 아래의 인덱스 값이 있으므로 setSelectIndex의 값을 이전 값에서 +1 해준다.

**Comment by jun-05** [comment]
interface를 쓰시면 interface폴더내부에 interface 파일을 만드는게 더 좋지 않았나 싶었습니다.
